### PR TITLE
Range slice full monty

### PR DESCRIPTION
**Pull request by pramasoul**
Brings congruence with Cpython range slicing, at a cost of 204 bytes of text seg in `stmhal`. See the new `builtin_range_slice.py` test for example cases which fail without this. Configured off by default in `mpconfig.h` and on for `unix` and `stmhal` ports.

### CODE REVIEW

**Comment by dpgeorge** [comment]
This one will take some thought.  It's not exactly clear what the errors are (seems like there are a few independent ones).

Would this fix be applicable to all uses of mp_seq_get_fast_slice_indexes?  Ie replace mp_seq_get_fast_slice_indexes with your non-fast version?

It seems that at least there could be a check for 0 step.  Otherwise, what class of slicing does the current code get wrong?

File: tests/basics/builtin_range_slice.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+try:
+    range(4)[::0]
+except ValueError:
+    print("ValueError")
+else:
+    print("SKIP")
+    import sys
+    sys.exit(0)
+
+print(range(7,-2,-4)[2:-2:])
```
Comment (dpgeorge) [inline tests/basics/builtin_range_slice.py]: Note that, even though uPy gives the wrong range for this, it does give an empty range which is correct.  Ie, `list(range(7,-2,-4)[2:-2:])` is correct under uPy.

**Comment by pramasoul** [comment]
> what class of slicing does the current code get wrong?

The current code is in error for slices of negative ```step```, as ```mp_seq_get_fast_slice_indexes``` enforces range [0 .. len] on ```start``` and ```stop```, and swaps them if necessary to obtain ```start <= stop```. The Cpython code, mimicked in  ```mp_seq_get_slice_indices```, for negative ```step``` canonicalizes ```start``` and ```stop``` to [-1 .. len-1] rather than [0 .. len], and further allows ```start > stop```.

Try e.g. ```range(4)[::-1]```.

>Would this fix be applicable to all uses of mp_seq_get_fast_slice_indexes? Ie replace mp_seq_get_fast_slice_indexes with your non-fast version?

I had hoped so, but that proved not to work, as the differing canonicalization is incompatible with the existing list and tuple slicing code. No surprise, as "only slices with step=1 (aka None) are supported" in that code at present. Work for the future.

>your non-fast version

It's a different function, but yes, it is slower, by as much as 190ns (32 clock cycles). I'll see what I can do.

**Comment by dpgeorge** [comment]
Lets not worry about speed at this point.  It looks like the missing functionality is (at the moment) just confined to range, so I'd prefer to have a patch that was just within the range code.  Eg maybe just a special case for negative step?

**Comment by pramasoul** [comment]
```mp_seq_get_fast_slice_indexes``` destroys information that is needed to slice with negative step. It appears that having its value doesn't spare us any of the work done in ```mp_seq_get_slice_indices``` to get to the correct result.

Micropython makes the reasonable descope over Cpython of generally dealing only in slice steps of 1. Having a complete & correct ability to slice sequences at least makes the workarounds less ugly.

I am happy to move the code out of ```sequence.c``` and into ```objrange.c``` if it sits better there. Please advise.

**Comment by dpgeorge** [comment]
This patch was actually implemented in a slightly different way some time ago in #2577, see 8908e505ce23ccd1f0ee49f3ffa80bf78cfaccfa.  And some tests from this PR were added in 2a8a564fbd48fba25c4876ff3d4317152a1d647c.

All the tests from this PR now pass, except for one (`print(range(7,-2,-4)[2:-2:])`) which passes when wrapped in `list()` due to the way uPy canonicalises the slice.  An explicit test for this was added in c7a0e1472da8e1b8e359bc26f838ab20bab00a8a.

Since this PR is no longer needed I'm closing it.

**Pull request closed without merging**